extensions/amp-analytics/0.1/events.js

```diff
@@ -447,21 +477,144 @@ export class TimerEventTracker extends EventTracker {
+      if (!!stopTracker) {
+        const stopTrackerBuilder = stopTracker.add.bind(stopTracker, context,
+            timerStop['on'], timerStop,
+            this.handleTimerToggle_.bind(this, timerId, eventType, listener),
+            trackerProvider);
+        this.trackers_[timerId][this.TIMER_PARAMS_.STOP_BUILDER] =
+            stopTrackerBuilder;
+      }
+
+    }
+    return () => {
+      this.removeTracker_(timerId);
+    };
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  generateTimerId_() {
+    return ++this.timerIdSequence_;
+  }
 
+  /**
+   * Toggles which listeners are active depending on timer state, so no race
+   * conditions can occur in the case where the timer starts and stops on the
+   * same event type from the same target.
+   * @param {number} timerId
+   * @param {string} eventType
+   * @param {function(!AnalyticsEvent)} listener
+   * @private
+   */
+  handleTimerToggle_(timerId, eventType, listener) {
+    const timerSpec = this.trackers_[timerId];
+    if (!!timerSpec[this.TIMER_PARAMS_.INTERVAL_ID]) {
+      // Stop timer and listen for start.
+      this.stopTimer_(timerId);
+      if (!!timerSpec[this.TIMER_PARAMS_.START_BUILDER]) {
+        const unlistenStart = timerSpec[this.TIMER_PARAMS_.START_BUILDER]();
+        this.trackers_[timerId][this.TIMER_PARAMS_.UNLISTEN_START] =
+            unlistenStart;
+      }
+    } else {
+      // Start timer and listen for stop.
+      this.startTimer_(timerId, eventType, listener);
+      if (!!timerSpec[this.TIMER_PARAMS_.STOP_BUILDER]) {
+        const unlistenStop =
+            timerSpec[this.TIMER_PARAMS_.STOP_BUILDER]();
+        this.trackers_[timerId][this.TIMER_PARAMS_.UNLISTEN_STOP] =
+            unlistenStop;
+      }
+    }
+  }
+
+  /**
+   * @param {number} timerId
+   * @param {string} eventType
+   * @param {function(!AnalyticsEvent)} listener
+   * @private
+   */
+  startTimer_(timerId, eventType, listener) {
+    const timerSpec = this.trackers_[timerId];
+    if (!!timerSpec[this.TIMER_PARAMS_.INTERVAL_ID]) {
+      return; // Timer already running.
+    }
     const win = this.root.ampdoc.win;
     const intervalId = win.setInterval(() => {
       listener(this.createEvent_(eventType));
-    }, interval * 1000);
-    this.trackers_.push(intervalId);
-    win.setTimeout(() => {
-      this.removeTracker_(intervalId);
-    }, maxTimerLength * 1000);
-    if (callImmediate) {
+    }, timerSpec[this.TIMER_PARAMS_.INTERVAL_LENGTH] * 1000);
+    this.trackers_[timerId][this.TIMER_PARAMS_.INTERVAL_ID] = intervalId;
+    if (!!timerSpec[this.TIMER_PARAMS_.UNLISTEN_START]) {
+      timerSpec[this.TIMER_PARAMS_.UNLISTEN_START]();
+      delete this.trackers_[timerId][this.TIMER_PARAMS_.UNLISTEN_START];
+    }
+    if (!this.isRestartableTimer_(timerId)) {
+      win.setTimeout(() => {
+        this.removeTracker_(timerId);
+      }, timerSpec[this.TIMER_PARAMS_.MAX_TIMER_LENGTH] * 1000);
```

Comment:
the `maxTimerLength` is not clear to me now. If the timer is restartable, are we going to ignore it? 
I actually think the concept of `canRestart` is not clear. Suppose I can use `startSpec` feature to only start my timer after some point. It's actually not restartable, and maybe we should still respect the `maxTimerLength` in such scenarios.